app/server/appsmith-server/src/main/java/com/appsmith/server/services/ApplicationPageServiceImpl.java

```diff
@@ -667,4 +667,43 @@ public void generateAndSetPagePolicies(Application application, PageDTO page) {
                 });
     }
 
+    @Override
+    public Mono<Application> reorderPage(String applicationId, String pageId, Integer order) {
+        return applicationService.findById(applicationId, MANAGE_APPLICATIONS)
+                .switchIfEmpty(Mono.error(new AppsmithException(AppsmithError.ACL_NO_RESOURCE_FOUND, FieldName.APPLICATION, applicationId)))
+                .flatMap(application -> {
+                    // Update the order in unpublished pages here, since this should only ever happen in edit mode.
+                    final List<ApplicationPage> pages = application.getPages();
+
+                    ApplicationPage foundPage = null;
+                    for (final ApplicationPage page : pages) {
+                        if (pageId.equals(page.getId())) {
+                            foundPage = page;
+                        }
+                    }
+                    //Set the order for the pages which has null values. At present we dont store the order and hence we need to update the values
+                    for (int i = 1; i <= pages.size(); i++) {
+                        ApplicationPage page = pages.get(i - 1);
+                        if (page.getOrder() == null) {
+                            page.setOrder(i);
+                        }
+                    }
+
+                    if(foundPage != null){
+                        //Increment the order of subsequent pages
+                        for (final ApplicationPage page : pages) {
+                            if (page.getOrder() < foundPage.getOrder()) {
```

Comment:
For pages that have the order number `>=` the given `order`, we should increment those right?

```suggestion
                            if (page.getOrder() >= order) {
```

For example, among pages ordered `1, 2, 3, 4`, if the page ordered at 4 is to be set to 2, that means the current 2 and 3 should be moved to 3 and 4 right?

Reply:
Let's consider the above example. What i am doing is that, First increase the order of the pages, which means `2-->3` and `3-->4`. Once this process is done, i will set the value of the page `foundPage` to the requested order. 

The idea is that suppose there 10 pages and the user re-ordered a page from 5--->2, we don't have to touch the order that is greater than 5 right ? Does it make sense ?

Reply:
Ah yes, good point. In that case, the condition is actually a bit more complicated. In your example, moving page from order 5 to order 2, that means `foundPage.getOrder()` is going to return 5. So in current implementation, we are incrementing _every_ page order that's less than 5, although, we shouldn't increment 1, since are moving to 2 right?

So, I think it should be something more like: `page.getOrder() >= foundPage.getOrder() && page.getOrder() < order`.

```suggestion
                            boolean isMovingUp = order < foundPage.getOrder();
                            if (isMovingUp ? (page.getOrder() >= order && page.getOrder() < foundPage.getOrder()) : (page.getOrder() < order && page.getOrder() >= foundPage.getOrder())) {
```

This will work when the page at 5 is moving to 2 among 10 pages, or when the page at 2 is moving to 5 among 10 pages. I wrote in a horribly unreadable way, please refactor it to use some local variables and also put in some comments or I'll have no idea what's happening there in a week or two. 😅

Also, let's add some tests to this logic please 😬